packages/core/src/utils/cursor.ts

```diff
@@ -0,0 +1,133 @@
+import type { AnyColumn, AnyColumnWithTable, OperandValueExpression, SelectQueryBuilder } from "kysely"
+import { z } from "zod"
+
+export interface PaginatedResult<T> {
+  data: T[]
+  next: Cursor | null
+}
+
+// Cursors are opaque strings that are parsed in the individual repositories
+export const CursorSchema = z.string()
+
+export const PaginateInputSchema = z
+  .object({
+    take: z.number(),
+    cursor: CursorSchema.optional().nullable(),
+  })
+  .optional()
+  .default({ take: 20, cursor: undefined })
+
+export type Pageable = z.infer<typeof PaginateInputSchema>
+export type Cursor = z.infer<typeof CursorSchema>
+
+type GetNextCursorOptions<T> = {
+  pageable: Pageable
+  buildCursor: (record: T) => Cursor
+}
+
+function getNextCursor<T>(records: T[], options: GetNextCursorOptions<T>): Cursor | null {
+  // We fetched take+1 records to determine if there is a record behind `take`.
+  // If we got exactly `take+1` records, there is a next page.
+  const hasNextPage = records.length === options.pageable.take + 1
+  if (!hasNextPage) {
+    return null
+  }
+
+  // biome-ignore lint/style/noNonNullAssertion: We know there is a record at -1 from the check above.
+  return options.buildCursor(records.at(-1)!)
+}
+
+type PaginatedQueryOptions<DB, TB extends keyof DB, C extends AnyColumn<DB, TB> | AnyColumnWithTable<DB, TB>, O> = {
+  pageable: Pageable
+  order: "asc" | "desc"
+  decodeCursor?: (cursor: Cursor) => OperandValueExpression<DB, TB, C>
+  buildCursor?: (record: O) => Cursor
```

Comment:
Should not be necessary, as far as I can tell cursor-encode/decode is only dependent on which columns you have

Reply:
Could you explain the use case of having individual sorting on each column? Are they not the same sorting order in 99% of the time?

Reply:
Hmm I guess if you have a table where you can sort on each column individually then it's useful.

Reply:
That functionality is not what I'm describing. Though it is used quite a few places in e.g. OW4's admin-panel, mostly useful when working directly towards a table.

What I'm talking about here is rather that instead of the users of this API having to specify how to encode/decode the cursor, that should be a part the implementation detail of our CursorPagination-package/module/whatever. That would make the call-locations more focused on intent, instead of technical implementation-details (I think you would call that a leaky abstraction)

This does reduce what you can implement on top of this, but I think it is better to extend the functionality of CursorPagination rather than making call-locations more complex, and solves the general problem rather than hacking stuff together with tight coupling to e.g. our primary key-usage

Reply:
Just fixed default cursor-encode/decode for multiple columns.

Reply:
With the current api you cannot order on individual columns when there are multiple columns spesified, as this makes things quite a bit more complex. What do you think about this @junlarsen?